src/app/components/subjects/subject-page/subject-detail/subject-detail.component.ts

```diff
@@ -0,0 +1,270 @@
+import { ActivatedRoute, Router } from '@angular/router';
+import { BsModalService, BsModalRef } from 'ngx-bootstrap/modal';
+
+import { ComponentCanDeactivate } from '../../../../guards/exit.subject-detail-page.guard';
+import { Observable } from "rxjs";
+
+import { Subject } from '../../../../common/entities/subject';
+import { Teacher } from '../../../../common/entities/teacher';
+import { Student } from '../../../../common/entities/student';
+
+import { StudentMark } from '../../../../common/entities/student-mark';
+import { Mark } from '../../../../common/entities/mark';
+
+import { ModalContentComponent } from '../../../../shared/components/modal-content/modal-content.component';
+import { ModalComponent } from '../../../../shared/components/modal/modal.component';
+
+@Component({
+  selector: 'app-subject-detail',
+  templateUrl: './subject-detail.component.html',
+  styleUrls: ['./subject-detail.component.scss']
+})
+export class SubjectDetailComponent implements OnInit, AfterViewChecked, ComponentCanDeactivate {
+  @ViewChild(ModalComponent, { static: false })
+  private templateModalComponent: ModalComponent;
+
+  saved: boolean = false;
+
+  bsModalRef: BsModalRef;
+  visibilitySaveButton: boolean = false;
+  buttonInfo: string = "Back to subject list";
+  markRegExp: any = /^[0-9]+$/;
+  dateRegExp: any = /^[0-9\/]+$/;
+
+  teacherTitle: string;
+  itemSelected: string = "";
+
+  subject: string;
+  idTeacher: string;
+  newIdTeacher: string;
+
+  studentsInfo: StudentMark[] = [];
+  subjectInfo: Subject;
+  teacher: Teacher;
+  students: Student[] = [];
+  dates: string[] = [];
+
+  constructor(
+    private dataService: DataService,
+    private activateRoute: ActivatedRoute,
+    private modalService: BsModalService,
+    private router: Router
+  ) {
+    this.idTeacher = activateRoute.snapshot.params['teacherId'];
+    this.newIdTeacher = activateRoute.snapshot.params['teacherId'];
+    this.subject = activateRoute.snapshot.params['id'];
+  }
+
+  ngOnInit() {
+    this.getTeacher(this.idTeacher);
+    this.getSubjectInfo(this.subject);
+    this.getStudentsById(this.studentsInfo);
+    this.getDates(this.studentsInfo);
+  }
+
+  ngAfterViewChecked() {
+    if (this.bsModalRef &&
+      this.itemSelected !== this.bsModalRef.content.itemSelected &&
+      this.bsModalRef.content.itemSelected) {
+      Promise.resolve(null).then((value) => {
+        let teacherSelected = this.bsModalRef.content.itemSelected;
+        this.itemSelected = teacherSelected;
+        this.teacherTitle = teacherSelected.split(" (id:")[0];
+        this.newIdTeacher = teacherSelected.split(" (id: ")[1];
+        this.newIdTeacher = this.newIdTeacher.split(")")[0];
+        this.visibilitySaveButton = true;
+      })
+    }
+  }
+
+  canDeactivate(): boolean | Observable<boolean> {
+
+    if (!this.saved && this.visibilitySaveButton) {
+      return confirm(
+        `If you leave the page without saving the changes, they will be lost.
```

Comment:
Why do you save some messages as constants (at constant.ts) and some writes inside the func?